dapr/proto/components/v1/state.proto

```diff
@@ -0,0 +1,241 @@
+  // Maximum of results that should be returned.
+  int64 limit = 1;
+  // The pagination token.
+  string token = 2;
+}
+
+message Query {
+  // Filters that should be applied.
+  map<string, google.protobuf.Any> filter = 1;
+  // The sort order.
+  repeated Sorting sort = 2;
+  // The query pagination params.
+  Pagination pagination = 3;
+}
+
+// QueryRequest is for querying state store.
+message QueryRequest {
+  // The query to be performed.
+  Query query = 1;
+  // Request associated metadata.
+  map<string, string> metadata = 2;
+}
+
+// QueryItem is an object representing a single entry in query results.
+message QueryItem {
+  // The returned item Key.
+  string key = 1;
+  // The returned item Data.
+  bytes data = 2;
+  // The returned item ETag
+  dapr.proto.common.v1.Etag etag = 3;
+  // The returned error string.
+  string error = 4;
+  // The returned contenttype
+  string content_type = 5;
+}
+
+// QueryResponse is the query response.
+message QueryResponse {
+  // The query response items.
+  repeated QueryItem items = 1;
+  // The response token.
+  string token = 2;
+  // Response associated metadata.
+  map<string, string> metadata = 3;
+}
+
+// TransactionalStateStore service provides a gRPC interface for transactional state store components.
+service TransactionalStateStore {
+    // Multi executes multiples operation in a transactional environment.
+    rpc Multi(TransactionalStateRequest) returns (TransactionalStateResponse) {}
+}
+
+// TransactionalStateOperation describes operation type, key, and value for transactional operation.
+message TransactionalStateOperation {
+  // request is either delete or set.
+  oneof request {
+    DeleteRequest delete = 1;
+    SetRequest set = 2;
+  }
+}
+
+// TransactionalStateRequest describes a transactional operation against a state store that comprises multiple types of operations
+// The Request field is either a DeleteRequest or SetRequest.
+message TransactionalStateRequest {
+  // Operations that should be performed.
+  repeated TransactionalStateOperation operations = 1;
+  // Request associated metadata.
+  map<string, string> metadata = 2;
+}
+
+// reserved for future-proof extensibility
+message TransactionalStateResponse {}
+
+// StateStore service provides a gRPC interface for state store components.
+service StateStore {
+  // Initializes the state store component with the given metadata.
+  rpc Init(InitRequest) returns (InitResponse) {}
+
+  // Returns a list of implemented state store features.
+  rpc Features(FeaturesRequest) returns (FeaturesResponse) {}
+
+  // Deletes the specified key from the state store.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Get data from the given key.
+  rpc Get(GetRequest) returns (GetResponse) {}
+
+  // Sets the value of the specified key.
+  rpc Set(SetRequest) returns (SetResponse) {}
+
+  // Ping the state store. Used for liveness porpuses.
+  rpc Ping(PingRequest) returns (PingResponse) {}
```

Comment:
Something I have been wondering: what exactly is the semantics of this method? Is it a "component is up and responding" or does it work more like a deep ping, reaching out to the underlying data store it is connecting Dapr to? Should it return a boolen or error status?

Suppose I am a component that provides connectivity to a ElasticSearch cluster. The cluster is unreachable. What should `ping` implementation do in this case?

1.  Raise an exception?
2. Nothing -- the component is up, that the cluster is unreachable is another concern. 
3. Ping should be a blocking call down to the downstream service, so... unreachability means it times out?

Reply:
Liveness in this case means: I can reach the component and he's responding to a Ping request. It is exactly like `ping` for testing connectivity over ICMP.

However, I agree with you that we should define the exactly associated semantics that we expect from the component, for instance:

If I can reach the component but some dependency is down, should it return an OK PingResponse? 

I think we can have different probes for it in the next releases like readiness/liveness, or even better - kubernetes might solve this for us.

Thanks for addressing this.

Reply:
> Liveness in this case means: I can reach the component and he's responding to a Ping request. It is exactly like ping for testing connectivity over ICMP.

True. As long as that is the semantic we are aiming at, which is different than "TCP port is open" which is also different than and "I got a 200 OK HTTP response".  In my head I thought ping was a "deep ping" but.. my bad. :)

Reply:
Some components do offer a deep check, like `select 1 from dual;`.